core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -521,402 +185,92 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
-   */
-  private void seekInternalWithCachingPartiallyReadBlock(long pos) throws IOException {
-    // Precompute this because mPos will be updated several times in this function.
-    final boolean isInCurrentBlock = mCurrentBlockInStream != null
-        && mPositionState.isInCurrentBlock(pos);
-
-    if (isInCurrentBlock && isReadFromLocalWorker()) {
-      // no need to partial cache the current block, and the seek is within the block
-      // so directly seeks to position.
-      mPositionState.setPos(pos);
-      mCurrentBlockInStream.seek(mPositionState.getPos() % mBlockSize);
-      return;
-    }
-
-    // cache the current block if neither of these conditions hold:
-    // (1) this is the first seek before any read, and the seek is outside the first block
-    // (2) the in stream reads from the local worker
-    // (3) the in stream reads from a remote worker but there is no local worker
-    boolean firstSeekOutsideFirstBlock =
-        mPositionState.getPos() == 0 && mCurrentBlockInStream == null && !isInCurrentBlock;
-    if (!firstSeekOutsideFirstBlock && canCacheToLocalWorker()) {
-      // Make sure that mCurrentBlockInStream and mCurrentCacheStream is updated.
-      // mPos is not updated here.
-      updateStreams();
 
-      // Cache till pos if seeking forward within the current block. Otherwise cache the whole
-      // block.
-      if (isInCurrentBlock && pos > mPositionState.getPos()) {
-        readCurrentBlockToPos(pos);
-      } else {
-        readCurrentBlockToEnd();
-      }
-
-      // Early return if we are at pos already. This happens if we seek forward with caching
-      // enabled for this block.
-      if (mPositionState.getPos() == pos) {
-        return;
+    if (mBlockInStream != null && mBlockInStream.remaining() == 0) { // current stream is done
+      closeBlockInStream();
+    }
+
+    /* Create a new stream to read from mPosition. */
+    // Calculate block id.
+    long blockId = mStatus.getBlockIds().get(Math.toIntExact(mPosition / mBlockSize));
+    // Create stream
+    mBlockInStream = mBlockStore.getInStream(mOptions.getBlockInfo(blockId), mOptions);
+    // Set the stream to the correct position.
+    long offset = mPosition % mBlockSize;
+    mBlockInStream.seek(offset);
+  }
+
+  private void closeBlockInStream() throws IOException {
+    if (mBlockInStream != null) {
+      // Clean up the current stream.
+      WorkerNetAddress dataSource = mBlockInStream.getAddress();
+      long blockId = mBlockInStream.getId();
+      mBlockInStream.close();
+      mBlockInStream = null;
+
+      // Send an async cache request to a worker based on partial and passive cache options.
+      boolean passiveCache = Configuration.getBoolean(PropertyKey.USER_FILE_PASSIVE_CACHE_ENABLED);
+      boolean partialCache = mOptions.getOptions().getCachePartiallyReadBlock();
+      long channelTimeout = Configuration.getMs(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+      if (partialCache) {
```

Comment:
Q: when `!partialCache`, if reading an entire block, we should still cache the block ,right? I could not find the logic.

Reply:
Good point, this is one of the downsides of using an RPC for async cache (as opposed to piggybacking off the read request). The case we do not catch is if passive caching is on, partial caching is off, and we read a full block. 

The condition of caching if the full block is read is an artifact of our previous implementation without much logical value. I think we should rename or remove partial caching and just use the read type.